### DISCUSSION TITLE
Flexible PID via Home Assistant

### DISCUSSION THREAD

**Discussion by JeanOlivier** [revk/xESP32-Faikout#177] [Show and tell]
The auto mode included in the firmware is very convenient, but maybe like me you'd like to have more flexibility.
e.g. have easy to adjust and familiar P, I, D parameters to fine-tune your heat pump to your specific situation, use another thermometer as the reference, etc.

That's why I wrote a prototype python PID script that uses pyscript from HACS to control my Daikin target temperature.
This script uses the value of the ble sensor as its temperature target.

It's tedious to adjust the PID parameters, but once tuned it oscillates slowly around the target temperature and reacts to sudden changes very well (sun showing up, oven turning on, etc.).

See here with a target of 22.5 on the ble sensor.

![image](https://github.com/revk/ESP32-Faikin/assets/11236668/359b4d1c-bc0d-4fcc-8eee-c492b1edd696)

I'm sharing here in the hope that it might be of use to others.

See https://hacs-pyscript.readthedocs.io/en/latest/installation.html to install and configure pyscript. Don't forget to reboot the system that Home Assistant is running on afterwards.

You can find my script here: https://gist.github.com/JeanOlivier/08ebf167e77b14a6b28f1d75bd67bb18#file-daikin_pid-py

It should be fairly self explanatory if you understand python. Just configure the header according to your needs.

**Comment by revk** [comment]
That is really interesting. I have achieved similarly tight controls as well. The problem I found was not so much the feedback controls but the fact that I am second guessing the operation of the a/c unit itself. Generally it was just turning the compressor on or off (well, seemed to be in my case) so setting a more fine tuned target temperature did not really control what it did. This meant the output of my control was either "high" or "low" enough to make it turn the compressor on or off. If I had any way to control the actual compessor directly I am sure I could have done even better. So impressed you managed it, well done.

**Reply by JeanOlivier** [reply]
Thanks! 

It was tedious to adjust, especially the derivative term. But once setup properly, it basically starts the compressor for about 10 minutes every half hour, with the temperature at the ble slowly oscillating with the same period at an amplitude of about ±0.2°C around target.

I did manage to tune it so that it was even more stable, but the compressor was turning on/off too quickly for my taste. No doubt that we could both do something much better if we could control the compressor state/rpm directly.

My goal in writing this was also to allow me to transparently switch between the heat pump and regular electric heating (via smart thermostat) if it's too cold outside. The heat pump can't keep up so well under -15°C. That's not implemented yet, as I just received the thermostats.


It's still a prototype, i.e. not clean nor integrated into HA as a virtual device, but it's been heating my house for a few months already so I thought I might as well share it, especially since people were asking about using the ble as a target. 
If everything is perfectly adjusted and the temperatures are stable, this code will stabilize the target of the Daikin to the correct delta T in order to reach the ble target; effectively replacing the internal sensor.

**Reply by turbolooser** [reply]
could you share how this could work for 3 indoor units using the same outdoor unit?
How to make every of the 3 scripts aware of the device to run on in HA?

**Reply by 44616E** [reply]
I'm also working on a similar idea, but have 6x 9.5kw air cconditioners in the same room so controlling all of them together and also keeping it in the power budget. Have 2 up and running now just waiting for my next order of faikins to get delivered to parcel forwarder and then to me.

@revk off topic but why did you select for amazon to not ship internationally, its not a big deal just curious?

**Reply by 44616E** [reply]
@JeanOlivier @revk What about using demand control to lower the compressor output so you can run it longer at lower duty?

**Reply by revk** [reply]
I have not "selected Amazon not to ship internationally". Getting them to do so is not simple, it seems they *are* doing so for a lot of countries where Amazon buy from me and sell internationally (which I did select). Trying to sell as me in EU is a lot of hurdles. We have customers all over the world, via Amazon, already. Just very unclear exactly where and when this works.

**Reply by 44616E** [reply]
@revk After our amazon order being stuck on ordered and not shipping for more than a week. I cancelled it and reordered via buy it again and it allowed me to ship direct to Australia. Weird it doesnt allow that through the normal button, hopefully they actually send it.

**Reply by revk** [reply]
Best of luck. Amazon are strange.

**Comment by JeanOlivier** [comment]
> @JeanOlivier @revk What about using demand control to lower the compressor output so you can run it longer at lower duty?

I think it's a good idea, but it wasn't supported when I wrote this. I'm also not sure if my own unit supports it correctly. IMO it'd be simpler to set the demand either manually or via a simple function of the temperature, then let the PID do its thing.

I might look into this eventually, but no guarantee. Feel free to experiment on your side and share the results.


Edit: if I were you, I'd try to run each unit via their own script with their own PID parameters, but set the same overall target. Thai way at equilibrium, I think each unit would contribute optimally to the total heating according to how effective they are.

**Reply by 44616E** [reply]
Some experimenting to be done, be interesting to see if they are more efficent at a specific demand, hard to determine that tho.

Power consumption is more critical then maintaining temperature. In my application.

**Comment by jsapede** [comment]
very interesting, but quite noob at python. how do you change the target temperature of 22.5 ?
would it be possible to integrate your algorythm in a standard climate entity ? or to interact with such climate entity to recieve the target temperature ?

**Reply by JeanOlivier** [reply]
You mean you're new to Python? Don't worry it's an easy language to read and adapt.

Currently you'd change de target by updating the script. This is a very preliminary sketch, hence the gist format, but it works. I did not have the time to implement a "virtual device" to control it further, but that was the plan. For now it's "good enough" for my needs though, so don't hold your breath.
